File: 字节跳动高频/112.PathSum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    bool dfs(TreeNode* root, int curSum, int sum) {
        if (root == NULL) return false;

        curSum += root->val;
        if (root->left == NULL && root->right == NULL && curSum == sum) return true; 
        return dfs(root->left, curSum, sum) || dfs(root->right, curSum, sum);
    }
public:
    bool hasPathSum(TreeNode* root, int sum) {
        return dfs(root, 0, sum);
    }
};

File: 字节跳动高频/112.PathSum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def dfs(self, root: TreeNode, curSum: int, sum: int) ->bool:
        if root == None:
            return False

        curSum += root.val
        if root.left == None and root.right == None and curSum == sum:
            return True
        
        return self.dfs(root.left, curSum, sum) or self.dfs(root.right, curSum, sum)

    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        return self.dfs(root, 0, sum)
